polytope/solvers.py

```diff
@@ -145,6 +158,34 @@ def _solve_lp_using_scipy(c, G, h):
         fun=sol.fun)
 
 
+def _solve_lp_using_gurobi(c, G, h):
+    """Attempt linear optimization using gurobipy"""
+    _assert_have_solver('gurobi')
+    m = gurobi.Model()
+    x = m.addMVar(G.shape[1], lb=-gurobi.GRB.INFINITY)
+    m.addConstr(G@x <= h)
+    m.setObjective(c@x)
+    m.optimize()
+
+    result = dict()
+    if m.Status == gurobi.GRB.OPTIMAL:
+        result['status'] = 0
+        result['x'] = x.x
+        result['fun'] = m.ObjVal
+        return result
+    elif m.Status == gurobi.GRB.INFEASIBLE:
+        result['status'] = 2
+    elif m.Status == gurobi.GRB.INF_OR_UNBD or m.Status == gurobi.GRB.UNBOUNDED:
```

Comment:
Good catch! There are at least 4 approaches here:
1. change `DualReductions` to 0, call `logger.warning`, and run again;
2. `result['status'] = 4 if m.Status == gurobi.GRB.INF_OR_UNBD` and add a comment about `DualReductions` and manually using `gurobipy` in `result['message']`;
3. same as previous, but add `**kwargs` to `lpsolve()`, and then forward these keyword arguments to `_solve_lp_using_gurobi`, handling an optional `dual_reductions` keyword;
4. similar to previous, but create a special solver name `"gurobi_DualReductions"` that can be used to run with `DualReductions=0`.

Reply:
I found that in my local installation of `gurobipy`, the status code `INF_OR_UNBD` occurs for an optimization problem with objective `min x` and feasible set `x =< 1`, so I added it to the function test_gurobipy_return_same_result_as_scipy(), confirming that the result status is unbounded and warning if INF_OR_UNBD did not occur. This is only a warning because it is an implementation detail where, if Gurobi identifies the problem as unbounded on the first run, the optimization result is correct. Warning encourages us to investigate why reoptimize is not necessary, but doing so is not urgent.

SciPy linprog may [include a field `message` in the optimization result](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.linprog.html). We are not currently using it explicitly in `polytope.solvers`, so in the case of reoptimizing with Gurobi, I added the note:
```py
result['message'] = 'Gurobi optimization status was INF_OR_UNBD, so reoptimized with DualReductions=0'
```

Reply:
Thanks for the discussion. I implemented automatic rerunning with DualReductions=0 in https://github.com/tulip-control/polytope/pull/98/commits/5753984e00896e314ecc899e8823b3f5503065b2